sql/src/main/java/io/crate/operation/projectors/ShardingUpsertExecutor.java

```diff
@@ -255,7 +253,7 @@ private void addToPendingRequests(TItem item, String indexName) {
 
     private void sendRequestsForBulk(CompletableFuture<BitSet> bulkResultFuture, Map<ShardLocation, TReq> bulkRequests) {
         if (bulkRequests.isEmpty()) {
-            bulkResultFuture.complete(responses);
+            bulkResultFuture.complete(null);
```

Comment:
the argument is never used, so I changed it to `null`

Reply:
I'm not 100% sure but I think `sendRequestsForBulk` is executed multiple times for each bulk. `responses` can be changed by one of the Futures. So changing this would not be safe.

Reply:
don't understand why this would not be safe. `responses` is never used in the consumer:

BatchIteratorBackpressureExectutor:

```java
private void continueConsumptionOrFinish(Object ignored, Throwable failure) { ... }
```

Reply:
The underlying implementation might change. Isn't the contract of a CompletableFuture to fulfill `result != null` XOR `failure != null`? Passing a reference and ignoring it is not really expensive.

Reply:
Got it. It's just that the current implementation is quite confusing.